File: csharp/core/PacketRegistry.cs
using com.github.fernthedev.lightchat.core.packets;
using com.github.fernthedev.lightchat.core.util;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using System.Text;

namespace com.github.fernthedev.lightchat.core
{
    public class PacketRegistry
    {
        public enum RegisteredReturnValues
        {
            NOT_IN_REGISTRY,
            IN_REGISTRY,
            IN_REGISTRY_DIFFERENT_PACKET
        }

        private PacketRegistry() { }

        private static readonly Dictionary<string, Type> PACKET_REGISTRY = new Dictionary<string, Type>();

        public static Type getPacketClassFromRegistry(string name)
        {
            if (!PACKET_REGISTRY.ContainsKey(name)) throw new InvalidOperationException("The packet registry does not contain packet \"" + name + "\" in the registry. Make sure it is spelled correctly and is case-sensitive.");

            return PACKET_REGISTRY[name];
        }

        public static GenericType<Packet> registerPacket(Packet packet)
        {
            if (PACKET_REGISTRY.ContainsKey(packet._PacketName) && PACKET_REGISTRY[packet._PacketName] != packet.GetType()) throw new InvalidOperationException("The packet " + packet.GetType().FullName + " tried to use packet name \"" + packet._PacketName + "\" which is already taken by the packet " + getPacketClassFromRegistry(packet._PacketName));

            PACKET_REGISTRY.Add(packet._PacketName, packet.GetType());

            return new GenericType<Packet>(packet);
        }

        public static Type registerPacket(Type packet)
        {
            var name = Packet.getPacketName(packet);

            if (PACKET_REGISTRY.ContainsKey(name) && PACKET_REGISTRY[name] != packet) throw new ArgumentException("The packet name \"" + name + "\" is already taken by the packet " + getPacketClassFromRegistry(name));

            PACKET_REGISTRY.Add(name, packet);

            return packet;
        }

        public static RegisteredReturnValues checkIfRegistered(Packet packet)
        {
            if (packet == null || !PACKET_REGISTRY.ContainsKey(packet._PacketName)) return RegisteredReturnValues.NOT_IN_REGISTRY;

            return PACKET_REGISTRY[packet._PacketName] == packet.GetType() ? RegisteredReturnValues.IN_REGISTRY : RegisteredReturnValues.IN_REGISTRY_DIFFERENT_PACKET;
        }

        public static RegisteredReturnValues checkIfRegistered(string packet)
        {
            if (packet == null || !PACKET_REGISTRY.ContainsKey(packet)) return RegisteredReturnValues.NOT_IN_REGISTRY;

            return RegisteredReturnValues.IN_REGISTRY;
        }



        private static IEnumerable<Type> GetTypesInNamespace(string nameSpace)
        {
            return GetTypesInNamespace(Assembly.GetExecutingAssembly(), nameSpace);
        }

            private static IEnumerable<Type> GetTypesInNamespace(Assembly assembly, string nameSpace)
        {
            return
              assembly.GetTypes()
                      .Where(t => t.Namespace != null && (t.Namespace.StartsWith(nameSpace, StringComparison.Ordinal) || t.Namespace.Equals(nameSpace, StringComparison.Ordinal)))
                      .ToArray();
        }

        public static void registerDefaultPackets()
        {

            registerPacketPackage(GetTypesInNamespace(StaticHandler.PACKET_NAMESPACE));
        }

        public static void registerPacketPackage(IEnumerable<Type> classes)
        {
            var packetClasses = classes.Where(t => typeof(Packet).IsAssignableFrom(t)).ToArray();

            foreach (var packetClass in packetClasses) {
                StaticHandler.Core.Logger.Debug("Registering the class {0}", packetClass.FullName);
                if (packetClass == typeof(Packet)) continue;

                registerPacket(packetClass);

            }
        }
    }
}

File: dart/core/lib/src/transport/packet_registry.dart
import 'dart:collection';

import '../packets/handshake_packets.dart';
import '../packets/latency_packets.dart';
import '../packets/other_packets.dart';
import '../packets/packets.dart';
import '../variables.dart';



class PacketRegistry {
  static final Map<String, Packet> packetRegistry = HashMap();

  static bool _defaultPacketRegistered = false;

  static Packet getPacketInstanceFromRegistry(
      String name, Map<String, dynamic> json) {
    if (!packetRegistry.containsKey(name)) {
      throw Exception('The packet registry does not contain packet $name in the registry. Make sure it is spelled correctly and is case-sensitive.');
    }

    return packetRegistry[name]!.fromJson(json);
  }

  static Type registerPacket(Packet packet) {
    if (packetRegistry.containsKey(packet.packetName) &&
        packetRegistry[packet.packetName] != packet) {
      throw ('The packet ${packet.runtimeType} tried to use packet name "${packet.packetName}" which is already taken by the packet ${packetRegistry[packet.packetName].runtimeType}');
    }

    packetRegistry[packet.packetName!] = packet;

    return packet.runtimeType;
  }

  static RegisteredReturnValues checkIfRegistered(Packet packet) {
    if (!packetRegistry.containsKey(packet.packetName)) {
      return RegisteredReturnValues.NOT_IN_REGISTRY;
    }

    return packetRegistry[packet.packetName].runtimeType == packet.runtimeType
        ? RegisteredReturnValues.IN_REGISTRY
        : RegisteredReturnValues.IN_REGISTRY_DIFFERENT_PACKET;
  }

  static RegisteredReturnValues checkIfRegisteredIdentifier(
      String packetIdentifier) {
    if (!packetRegistry.containsKey(packetIdentifier)) {
      return RegisteredReturnValues.NOT_IN_REGISTRY;
    }

    return RegisteredReturnValues.IN_REGISTRY;
  }

  static void registerDefaultPackets() {
    if (_defaultPacketRegistered) return;

    var packets = List<Packet>.of({
      InitialHandshakePacket.constant,
      KeyResponsePacket.constant,
      ConnectedPacket.constant,
      RequestConnectInfoPacket.constant,
      CommandPacket.constant,
      IllegalConnection.constant,
      MessagePacket.constant,
      SelfMessagePacket.constant,
      PingPacket.constant,
      PingReceive.constant,
      PongPacket.constant,
    });
    _defaultPacketRegistered = true;
    try {
      registerPackets(packets);
    } catch (e) {
      _defaultPacketRegistered = false;
      rethrow;
    }
  }

  static void registerPackets(List<Packet> packets) {
    for (var packet in packets) {
      if (Variables.debug) {
        print(
            'Registering ${packet.runtimeType} with name ${packet.packetName}');
      }

      registerPacket(packet);
    }

//    for (Package packageT : Package.getPackages()) {
//    if (packageT.getName().startsWith(StaticHandler.PACKET_PACKAGE)) {
//    StaticHandler.getCore().getLogger().debug("Registering the package " + packageT.getName());
//    registerPacketPackage(packageT.getName());
//    }
//    }
  }

//  static void registerPacketPackage(String packageName) {
//    Set<Class<? extends Packet>> classes = new Reflections(packageName).getSubTypesOf(Packet.class);
//
//    for (Class<? extends Packet> packetClass : classes) {
//    StaticHandler.getCore().getLogger().debug("Registering the class {}", packetClass);
//
//    try {
//    registerPacket(packetClass);
//    } catch (Exception e) {
//    e.printStackTrace();
//    }
//    }
//  }

}

enum RegisteredReturnValues {
  NOT_IN_REGISTRY,
  IN_REGISTRY,
  IN_REGISTRY_DIFFERENT_PACKET
}
